### ISSUE TITLE
proposal: Unified text primitives/nestable text elements

### ISSUE THREAD

**Issue by tropicaaal** [marc2332/freya#835]
Consider this a mini-RFC. I think I may have proposed it last year on discord, but it probably wasn't conveyed as well as it could have been.

# Freya's Text Primitives

Currently freya has three elements (primitives) related to rendering text:
- `label`
- `paragraph`
- `text`

These all render similar (or identical things), but serve slightly different purposes at the same time:
- `label` is generally used for "one-off" cases where you just need a single line of contigiously-styled text. It's simple and used for most things text-related where you just need one style. `label` cannot take children other than a raw string.
- `paragraph`/`text` are used when you need differently styled text nodes in the same text layout at once. For example, if you wanted **emphasized** text, you would use a paragraph/text combo.

```rs
paragraph {
    line_height: "2",
    text {
        font_size: "20",
        font_style: "normal",
        "Normal "
    }
    text {
        font_size: "20",
        font_style: "italic",
        "Italic "
    }
    text {
        font_size: "20",
        font_style: "oblique",
        "Oblique"
    }
}
```

## The case for a unified element

All three of these elements serve the same purpose -- drawing text, yet they have different arbitrary restrictions or API complexity. Currently, `paragraph` could almost certainly replace `label` in all cases, but that would be annoying to deal with due to having `paragraph` split into a pair of elements. Instead, what if there was a single simplified `text` primitive that served all the usecases. This would have some advantages:

- Freya wouldn't have to individually handle three elements in its native APIs, and all rendering and node-related code for text would be unified under a single primitive, reducing internal complexity.
- It's easier to teach to users. The difference between `label` and `paragraph` wouldn't need to be explained.
- It would remove any accidental differences in behavior or attributes between the elements as a result of feature additions to one element but not the other.
- It better fits the paradigm of having a set of "base elements". Generally most UI is visually comprised of text and rounded rectangles if you break it into its base parts, and this better fits that paradigm.

## A new `text` element

This proposal presents a new `text` element that takes the place of `paragraph`. In its simplest form, it looks a lot like the current `label`.
```rs
text {
    "Hi",
}
```

`text` elements can contain bare strings or `span`s as children.

```rs
text {
    "Hello",
    span {
        font_weight: "bold",
        "World"
    }
}
```

## The `span` element

`span` is basically a replacement for the previous `text` element, and defines a styled span of text in `SkParagraph`. Unlike the parent `text`, they can only take inline text styling properties such as `font_size`, `font_weight`, etc...

- A span that is not a child of `text` is meaningless and does not render anything at all.
- `span`s inherit the text styling properties of their parent `text`.
- A bare string child of `text` is equivalent to a `span` with no styles.

This reduces the number of text elements from 3 to 2 and provides a clear distinction between a `SkParagraph` and a text span in the paragraph. It also reduces the API complexity of using styled spanned text while also removing `label`.

## Past Work

This is very similar to the approaches of [React Native](https://reactnative.dev/docs/text) and HTML:

```html
<span>
    I am
    <span style="font-weight: bold;">
        bold
    </span>
</span>
```

Unlike HTML, there is a distinction between the text's *block* element (in Freya's case a `SkParagraph`) and the inline children.

## Remaining Questions
- Should `span`s be able to have `span` children to inherit styles? (probably not initially I guess...)

**Comment by marc2332** [comment]
Definitely an interesting suggestion @Tropix126.

Reducing internal complexity and simplifying the user-facing API is always welcome.

Now, I am not 100% sold.

Using the same element (`text`) for both the "parent" wrapper and inner text spans sounds like an over abstraction to me. They are in fact two different things if we look at the skia API.
`paragraph` and `text` for example have some key differences like `max_lines` and `text_align` (and maybe more APIs in the future) So, you cannot merge both elements (`paragraph` and `text`) into one.
You also cannot have infinite nested text nodes, unless the nested is only to have style inheritance, I say this because the SkParagraph API from Skia only supports 1 level text spans, so all these text spans would need to be flattened into an array.

I could maybeeee agree on merging `label` and `paragraph` (I'd need to think about it though), but `text` (as in text span) should still be its own thing, although it could be renamed.

- `label` + `paragraph` -> `label` or `text` or ? 
- `text` -> `text` or `span` or ?

**Comment by Tropix126** [comment]
> You also cannot have infinite nested text nodes, unless the nested is only to have style inheritance, I say this because the SkParagraph API from Skia only supports 1 level text spans, so all these text spans would need to be flattened into an array.

Yeah, in this case nesting would be for styling only. At the rendering level, all text spans would be flattened, with only the top level node responsible for rendering the actual paragraph. I'm not sure if I see that as a bad thing personally, just an implementation detail. It doesn't reflect the internal workings of the Skia API, but it also isn't taking away any control from users.

`max_lines` is an interesting case though, and I admittedly didn't consider that. Looking at CSS, the closest equivalent is the nonstandard `-webkit-max-lines` property, which is only valid on block elements in the layout and not inline elements (something that isn't distinguished in Skia).

I guess our options there with one element would be to only consider it on the top level `text`, although I agree that is a bit of an abstraction leak. The "block element" in this case is whatever the nearest non-text parent element is, so probably whatever container `rect` is holding the text. This would likely mean that `max_lines` and `text_align` would belong non-text elements rather than being an attribute of the text itself.

If you do want to have a simpler approach more reflective of the Skia APIs, then I guess two elements would be the way to go. I think `text`/`span` would work nicely for that.

**Comment by Tropix126** [comment]
I've updated the proposal to use two elements (one `text` block element and a `span` for inline styled text spans).
